Implement minimal SBI v0.3

ChinYikMing opened the pull request:
SBI acts as a communication layer between S-mode software and M-mode hardware. To boot Linux kernel, there are some minimal SBI extensions have to be implemented and they are:
1. Base extension(EID=0x10)
2. Timer extension(EID=0x54494D45)

SRST extension(EID=0x53525354) is optional so just implemented shutdown reason.

Related: #310

jserv reviewed (commented):
<!-- github-benchmark-action-comment(start): Benchmarks Summary -->
# Benchmarks

<details>

| Benchmark suite | Current: e9cecd11b0811b36af2a077743466d61235c7b6b | Previous: 487d14f57f947144eace29777fbabd8655cff50e | Ratio |
|-|-|-|-|
| `Dhrystone` | `1742.88` Average DMIPS over 10 runs | `1691` Average DMIPS over 10 runs | `0.97` |
| `Coremark` | `1499.792` Average iterations/sec over 10 runs | `1569.296` Average iterations/sec over 10 runs | `1.05` |

</details>

This comment was automatically generated by [workflow](https://github.com/sysprog21/rv32emu/actions?query=workflow%3ABenchmark) using [github-action-benchmark](https://github.com/marketplace/actions/continuous-benchmark).
<!-- github-benchmark-action-comment(end): Benchmarks Summary -->

jserv commented on src/riscv.h:
```diff
@@ -96,6 +96,28 @@ enum {
 #define MSTATUS_MPIE (1 << MSTATUS_MPIE_SHIFT)
 #define MSTATUS_MPP (3 << MSTATUS_MPP_SHIFT)
 
+#define SBI_SUCCESS 0
```
Add comments before making list of these SBI related macros. Something like SBI spec.

ChinYikMing committed:
Implement minimal SBI v0.3

SBI acts as a communication layer between S-mode software and M-mode
hardware. To boot Linux kernel, there are some minimal SBI extensions
have to be implemented and they are:
1. Base extension(EID=0x10)
2. Timer extension(EID=0x54494D45)

SRST extension(EID=0x53525354) is optional so just implemented shutdown
reason.

Related: #310

jserv commented on src/riscv.h:
```diff
@@ -96,6 +96,53 @@ enum {
 #define MSTATUS_MPP (3 << MSTATUS_MPP_SHIFT)
 
+/*
+ * SBI functions must return a pair of values:
```
Add some reference somewhere in this header.

jserv commented on src/syscall.c:
```diff
@@ -369,6 +372,96 @@ extern void syscall_setup_audio(riscv_t *rv);
 extern void syscall_control_audio(riscv_t *rv);
 #endif
 
+/*
```
Refine the comment as `/* SBI related system calls */` or
```c
/* SBI related system calls.
 * In case you need more lines.
 */
```

jserv commented on src/riscv.h:
```diff
@@ -96,6 +96,51 @@ enum {
 #define MSTATUS_MPP (3 << MSTATUS_MPP_SHIFT)
 
+/*
+ * SBI reference: https://github.com/riscv-non-isa/riscv-sbi-doc
```
Move the line to the end of this comment block.

jserv commented:
Thank @ChinYikMing for contributing!
